tribuo classification java example

snedamle opened the pull request.

snedamle committed:
tribuo classification java example

snedamle committed:
updated the conclave repo path to <path_to_conclave_sdk_repo>

snedamle committed:
removing un-necessary files.

rui-alm commented on tribuo-classification/client/build.gradle:
```diff
@@ -0,0 +1,17 @@
+    id 'java'
+    id 'application'
+}
+sourceCompatibility = 1.8
```
I believe the source compatibility is already being controlled by the `subprojects` section of `tribuo-classification/build.gradle`, so we should be able to remove this line.

snedamle replied:
done

rui-alm commented on tribuo-classification/client/src/main/java/com/r3/conclave/sample/client/Client.java:
```diff
@@ -0,0 +1,134 @@
+            MutableDataset trainingDataset = new MutableDataset<>(irisSplitter.getTrain());
+            MutableDataset testingDataset = new MutableDataset<>(irisSplitter.getTest());
+
+            System.out.println("helooooooo" + trainingDataset.getOutputs().size());
```
Remove or improve the logging.

snedamle replied:
done

rui-alm commented on tribuo-classification/client/src/main/java/com/r3/conclave/sample/client/Client.java:
```diff
@@ -0,0 +1,134 @@
+        EnclaveInstanceInfo instanceInfo = EnclaveInstanceInfo.deserialize(attestationBytes);
+
+//        //verify attestation received by enclave against the enclave code hash which we have
+//        EnclaveConstraint.parse("S:"+ constraint +" PROD:1 SEC:INSECURE" ).check(instanceInfo);
```
We need to reenable this. We can have the contraints being passed via command line arguments.

snedamle replied:
done

rui-alm commented on tribuo-classification/common/build.gradle:
```diff
@@ -0,0 +1,20 @@
+    id 'java'
+}
+
+sourceCompatibility = 1.8
+
+repositories {
+    mavenCentral()
+}
```
I believe the source compatibility is already being controlled by the `subprojects` section of `tribuo-classification/build.gradle`, so we should be able to remove these lines.

snedamle replied:
done

rui-alm commented on tribuo-classification/container-gradle:
```diff
@@ -0,0 +1,148 @@
+#!/usr/bin/env bash
```
This script is currently provided by the SDK, so it doesn't need to be in the project. We should remove it.

snedamle replied:
done

rui-alm commented on tribuo-classification/enclave/src/main/java/com/r3/conclave/sample/enclave/MlEnclave.java:
```diff
@@ -0,0 +1,100 @@
+            e.printStackTrace();
+        }
+        try {
+            return is.readObject();
```
I think we should merge these `try`s. If the `ObjectInputStream` initialization throws and exception, we'll get a `NullPointerException` here.

snedamle replied:
done

rui-alm replied:
Could you double check if it's done? I still see the 2 `try` statements separately.

rui-alm commented on tribuo-classification/enclave/src/test/java/com/r3/conclave/sample/enclave/MockTest.java:
```diff
@@ -0,0 +1,14 @@
+
+    @Test
+    void calculateAdConversionRate()  {
+
```
Are there any tests you'd like to add, or should this be removed?

snedamle replied:
added test case

rui-alm commented on tribuo-classification/host/build.gradle:
```diff
@@ -0,0 +1,31 @@
+    def graalVMDir = tasks.getByPath(":enclave:untarGraalVM").outputs.files.first()
+    def configurationDir = project(":enclave").projectDir.toPath().resolve("src/main/resources/META-INF/native-image")
+    def filterFile = configurationDir.resolve("filter.json")
+    application {
```
This needs to be outside of the `dependencies` block.

snedamle replied:
done

rui-alm commented on tribuo-classification/host/src/main/java/com/r3/conclave/sample/host/Host.java:
```diff
@@ -0,0 +1,96 @@
+
+            //forward the request to enclave. The request is encrypted using enclave's public
+            //key and hence is not accessible to host
+            enclaveHost.deliverMail(1, mailBytes, routingHint);
```
I think there should be a counter for each client. I'm guessing this works because we are assuming a single message per client? @shamsasari what do you think?

rui-alm replied:
Maybe I'm confusing this with some older API.

rui-alm commented on tribuo-classification/host/build.gradle:
```diff
@@ -0,0 +1,31 @@
+    def filterFile = configurationDir.resolve("filter.json")
+    application {
+        run.dependsOn(":enclave:untarGraalVM")
+        if (project.hasProperty("generateConfigFiles") && properties.get("enclaveMode").toString().toLowerCase() == "mock" != null) {
```
The `!= null` is a copy paste error from the `tribuo-tutorials`. It's not doing anything. We should remove it from the `tribuo-tutorials` as well.

snedamle replied:
done

rui-alm commented on tribuo-classification/client/src/main/java/com/r3/conclave/sample/client/Client.java:
```diff
@@ -0,0 +1,134 @@
+        }
+
+        //take inputs from user, populate the input data object with training, testing data and input type - train or evaluate
+        InputData inputData = new InputData();
```
I think we should modify the `InputData`'s constructors. The constructor with no arguments could take the `inputType` and the 3 argument constructor that takes the inputType and the datasets doesn't seem to be used.

snedamle replied:
done

rui-alm commented on tribuo-classification/enclave/src/main/java/com/r3/conclave/sample/enclave/MlEnclave.java:
```diff
@@ -0,0 +1,100 @@
+            //send evaluation results back to all the clients using the routingHintToPublicKey mapping
+            for (String key : routingHintToPublicKey.keySet()) {
+                byte[] encryptedReply = postOffice(routingHintToPublicKey.get(key)).
+                        encryptMail(evaluation.toString().getBytes(StandardCharsets.UTF_8));
```
I think its common style to put the `.` which invokes `encryptMail` on the same line as `encryptMail`.

snedamle replied:
done

rui-alm commented on tribuo-classification/client/src/main/java/com/r3/conclave/sample/client/Client.java:
```diff
@@ -0,0 +1,134 @@
+    /**
+     * Serialise the input data object containing training and testing data for sending it to enclave.
+     * @param obj
+     * @return
```
We should complete the javadoc for all the relevant code.

snedamle replied:
done

gendal commented on tribuo-classification/README.md:
```diff
@@ -0,0 +1,105 @@
+
+Start the host on a non-Linux system, which will build the enclave and host by default in simulation mode.
+
+    ./container-gradle host:run
```
if you delete container gradle per rui's recommendation, be sure to update this to something like `<PATH-TO-CONCLAVE-INSTALL-DIRECTORY>/scripts/container-gradle`

snedamle replied:
done

gendal commented on tribuo-classification/README.md:
```diff
@@ -0,0 +1,105 @@
+To read more on Conclave go to the documentation site - https://docs.conclave.net
+
+Please note the citation request made by the owners of the input data set used in this sample for testing.
+Citation Request:
```
is this compatible with the Apache 2 license I wonder?   We may need to add a clarifying comment in the README that the R3 _code_ is licensed under Apache 2 but the training data set has its own terms and conditions?

snedamle replied:
mentioned the same.

snedamle committed:
1. updated readme by mentioning R3 code is licensed under Apache 2
2. removed sourceCompatibility from build.gradle
3. added constraint to client code
4. removed unused constructors from InputData
5. refactored enclave code to be able to unit test the methods
6. added MockEnclave unit test
7. add data files for unit testing
8. moved the generateConfig task out of dependencies task

snedamle committed:
1. removing container-gradle script

rui-alm commented on tribuo-classification/client/src/main/java/com/r3/conclave/sample/client/Client.java:
```diff
@@ -0,0 +1,133 @@
+    public static void main(String[] args) throws InterruptedException, IOException, InvalidEnclaveException {
+
+        if (args.length == 0) {
+            System.err.println("Please pass [TRAIN/EVALUATE] [FILENAME] [CONSTRAINT] containing input dataset to train/test the model.");
```
I think the message is now confusing with the introduction of `[CONSTRAINT]`. The last part: "containing input dataset to train/test the model" refers to the filename. You might want to use `<paramName>`, which means mandatory parameter, rather than `[]`, which means optional.

rui-alm commented on tribuo-classification/client/src/main/java/com/r3/conclave/sample/client/Client.java:
```diff
@@ -0,0 +1,133 @@
+        EnclaveInstanceInfo instanceInfo = EnclaveInstanceInfo.deserialize(attestationBytes);
+
+//        //verify attestation received by enclave against the enclave code hash which we have
+        EnclaveConstraint.parse("S:"+ constraint +" PROD:1 SEC:INSECURE" ).check(instanceInfo);
```
The enclave code [constraint](https://docs.conclave.net/writing-hello-world.html#constraints) is specified using `C:<MRENCLAVE>`, rather than `S:<MRSIGNER>`.

snedamle committed:
1. use code signer constraint instead of code hash.
2. add constraint to client input parameters.
3. remove multiple try blocks, club them in 1 try catch block.

rui-alm commented on tribuo-classification/common/build.gradle:
```diff
@@ -0,0 +1,20 @@
+    id 'java'
+}
+
+sourceCompatibility = 1.8
+
+repositories {
+    mavenCentral()
+}
```
```suggestion
```

snedamle replied:
Hi @rui-alm what change have you introduced here? I cannot see any difference.

rui-alm replied:
It's removing the `sourceCompatibility` and the `repositories`. They are set on the `tribuo-classification/build.gradle`, aren't they?

rui-alm commented on tribuo-classification/enclave/src/main/java/com/r3/conclave/sample/enclave/MlEnclave.java:
```diff
@@ -0,0 +1,114 @@
+        //send evaluation results back to all the clients using the routingHintToPublicKey mapping
+        for (String key : routingHintToPublicKey.keySet()) {
+            byte[] encryptedReply = postOffice(routingHintToPublicKey.get(key)).
+                    encryptMail(evaluation.toString().getBytes(StandardCharsets.UTF_8));
```
```suggestion
            byte[] encryptedReply = postOffice(routingHintToPublicKey.get(key))
                    .encryptMail(evaluation.toString().getBytes(StandardCharsets.UTF_8));
```

rui-alm commented on tribuo-classification/enclave/build.gradle:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    implementation "org.slf4j:slf4j-api:1.7.30"
+    runtimeOnly "org.slf4j:slf4j-simple:1.7.30"
```
```suggestion
    implementation "org.slf4j:slf4j-api:$slf4j_version"
    runtimeOnly "org.slf4j:slf4j-simple:$slf4j_version"
```

rui-alm commented on tribuo-classification/gradle.properties:
```diff
@@ -0,0 +1,5 @@
+#Tue Feb 09 22:52:29 GMT 2021
+conclaveVersion=1.1
+conclaveRepo=<path_to_conclave_sdk_repo>
+slf4j_version = '1.7.30'
```
```suggestion
slf4j_version=1.7.30
```

snedamle committed:
1. removed source_compatibility=1.8 from enclave build.gradle as it is present in the main project's build.gradle
2. updated sl4j version property
